### ISSUE TITLE
Atrous/Dilated Convolutions cannot dynamically Compute Shape on TPU

### ISSUE THREAD

**Issue by hyang0129** [tensorflow/tensorflow#48602]
**System information**
- Colab Pro TPU Environment
- TF 2.4.1
- TPU v2-8

**Describe the current behavior**
tf.keras.layers.Conv2D(filters = 32, kernel_size=3, strides = 1, use_bias = False, dilation_rate = 2) does not compile on TPU. 

Building a custom layer using tf.nn.atrous_conv2d results in the same error. 

**Describe the expected behavior**
tf.nn.atrous_conv2d is a supported op based on this [list](https://cloud.google.com/tpu/docs/tensorflow-ops) of supported ops. 

**Standalone code to reproduce the issue**
Below notebook builds minimal model and trains on mnist. Works when dilation rate is not used for the conv2d. Otherwise generates the error. 

[colab notebook](https://colab.research.google.com/drive/1Hl5IZ9rnK6g58j0MSC3rGqTwrsXkrfHN?usp=sharing)

**Error code from Notebook**
```
InvalidArgumentError: 9 root error(s) found.
  (0) Invalid argument: {{function_node __inference_train_function_16413}} Compilation failure: Reshape's input dynamic dimension is decomposed into multiple output dynamic dimensions, but the constraint is ambiguous and XLA can't infer the output dimension %reshape.107 = f32[2,2,32,12,12,32]{5,4,3,2,1,0} reshape(f32[<=128,12,12,32]{3,2,1,0} %convolution.106), metadata={op_type="BatchToSpaceND" op_name="sequential_5/conv2d_10/Conv2D/BatchToSpaceND"}. 
	TPU compilation failed
	 [[{{node tpu_compile_succeeded_assert/_7203248111506234482/_5}}]]
	 [[TPUReplicate/_compile/_757783574996934175/_4/_300]]
  (1) Invalid argument: {{function_node __inference_train_function_16413}} Compilation failure: Reshape's input dynamic dimension is decomposed into multiple output dynamic dimensions, but the constraint is ambiguous and XLA can't infer the output dimension %reshape.107 = f32[2,2,32,12,12,32]{5,4,3,2,1,0} reshape(f32[<=128,12,12,32]{3,2,1,0} %convolution.106), metadata={op_type="BatchToSpaceND" op_name="sequential_5/conv2d_10/Conv2D/BatchToSpaceND"}. 
	TPU compilation failed
	 [[{{node tpu_compile_succeeded_assert/_7203248111506234482/_5}}]]
	 [[TPUReplicate/_compile/_757783574996934175/_4/_336]]
  (2) Invalid argument: {{function_node __inference_train_function_16413}} Compilation failure: Reshape's input dynamic dimension is decomposed into multiple output dynamic dimensions, but the constraint is ambiguous and XLA can't infer the output dimension %reshape.107 = f32[2,2,32,12,12,32]{5,4,3,2,1,0} reshape(f32[<=128,12,12,32]{3,2,1,0} %convolution.106), metadata={op_type="BatchToSpaceND" op_name="sequential_5/conv2d_10/Conv2D/BatchToSpaceND"}. 
	TPU compilation failed
	 [[{{node tpu_compile_succeeded_assert/_7203248111506234482/_5}}]]
	 [[tpu_compile_succeeded_assert/_7203248111506234482/_5/_279]]
  (3) Invalid argument: {{function_node __inference_train_function_16413}} Compilation failure: Reshape's input dynamic dimension is decomposed into multiple output dynamic dimensions, but the constraint is ambiguous and XLA can't infer the output dimension %reshape.107 = f32[2,2,32,12,12,32]{5,4,3,2,1,0} reshape(f32[<=128,12,12,32]{3,2,1,0} %convolution.106), metadata={op_type="BatchToSpaceND" op_name="sequential_5/conv2d_10/Conv2D/BatchToSpaceND"}. 
	TPU compilation failed
	 [[{{node tpu_compile_succeeded_assert/_7203248111506234482/_5}}]]
	 [[tpu_compile_succeeded_assert/_7203248111506234482/_5/_267]]
  (4) Invalid argument: {{function_node __inference_train_function_16413}} Compilation failure: Reshape's input dynamic dimension is decomposed into multiple output dynamic dimensions, but the constraint is ambiguous and XLA can't infer the output dimension %reshape.107 = f32[2,2,32,12,12,32]{5,4,3,2,1,0} reshape(f32[<=128,12,12,32]{3,2,1,0} %convolution.106), metadata={op_type="BatchToSpaceND" op_name="sequential_5/conv2d_10/Conv2D/BatchToSpaceND"}. 
	TPU compilation failed
	 [[{{node tpu_compile_succeeded_assert/_7203248111506234482/_5}}]]
	 [[tpu_compile_succeeded_assert/_7203248111506234482/_5/_291]]
  (5) ... [truncated]
```

**Comment by UsharaniPagadala** [comment]
@hyang0129 

Since similar issue is already being tracking in issues [#1110](https://github.com/tensorpack/tensorpack/issues/1110) [,#28788](https://github.com/tensorflow/tensorflow/issues/28788), [#26797](https://github.com/tensorflow/tensorflow/issues/26797), [#29542](https://github.com/tensorflow/tensorflow/issues/29542) .Could you please let us know if it helps.

To avoid duplicates can you please close this and subscribe/follow that issues? Thanks!

**Comment by hyang0129** [comment]
> @hyang0129
> 
> Since similar issue is already being tracking in issues [#1110](https://github.com/tensorpack/tensorpack/issues/1110) [,#28788](https://github.com/tensorflow/tensorflow/issues/28788), [#26797](https://github.com/tensorflow/tensorflow/issues/26797), [#29542](https://github.com/tensorflow/tensorflow/issues/29542) .Could you please let us know if it helps.
> 
> To avoid duplicates can you please close this and subscribe/follow that issues? Thanks!

This is not a duplicate of the issues you mentioned. Those issues deal with GPU hardware accelerators. This issue deals with TPU hardware accelerators. 

You can reproduce the issue by running the colab notebook posted above, using TF 2.4.1 and a TPU v2-8. 

This issue has not been resolved for TPU systems.

**Comment by hyang0129** [comment]
@UsharaniPagadala this is not a duplicate issue. This is specific to TPU hardware accelerators. Also, those issues that you listed have been resolved, but we still get the error above in TF 2.4.1 on a TPU.

**Comment by amahendrakar** [comment]
@ymodak,
I was able to reproduce the error with [TF v2.4.1](https://colab.research.google.com/gist/amahendrakar/1bd46aca07368f2057087de77774cbe7/48602.ipynb#scrollTo=CXtFIzZ0zD9s). 

Whereas, with [TF v2.5.0rc1](https://colab.research.google.com/gist/amahendrakar/6fc93724ae3b0f5836684ecbe8ad308e/48602-2-5.ipynb#scrollTo=CXtFIzZ0zD9s) and TF-nightly, the error changes to 
```
NotFoundError: Op type not registered 'XlaSetDynamicDimensionSize' in binary running on n-70c139a5-w-0. Make sure the Op and Kernel are registered in the binary running in this process. Note that if you are loading a saved graph which used ops from tf.contrib, accessing (e.g.) `tf.contrib.resampler` should be done before importing the graph, as contrib ops are lazily registered when the module is first accessed.
```

Please check the linked gist for reference. Thanks!

**Comment by hyang0129** [comment]
If anyone has a quick fix for this please let me know ASAP. I've got a deadline to hit and the latest I could start training on imagenet and get results in time would be end of April. For now, I'm substituting the dilated convs with non dilated convs.

**Comment by hthu** [comment]
Sorry for the slow response.

The root issue is that your dataset doesn't have static batch shape, and that leads to dynamic shape inside TPU XLA which is problematic.

You can simply use `drop_remainder=True` when calling the batch function to produce a `fixed` shaped batch, e.g.:

```python
data = data.batch(256, drop_remainder=True)
```

And that would give you the output shape below:
```
<MapDataset shapes: ((256, 28, 28, 1), (256,)), types: (tf.float32, tf.int64)>
```

With that, you should be able to train properly.

**Comment by hyang0129** [comment]
oh. That is... not anywhere in the documentation that I could find.

Can this be added somewhere or can we update the error generated when calling batch_to_space?

**Comment by hthu** [comment]
https://github.com/tensorflow/tensorflow/commit/6b746d6631b9dce36303ecac26b70fef9a9d8cea

The commit should make the documentation on dataset batching a bit more clear.
